sw/device/silicon_creator/rom_ext/rom_ext.c

```diff
@@ -257,18 +258,112 @@ void rom_ext_sram_exec(owner_sram_exec_mode_t mode) {
   }
 }
 
+// A version of spx_verify that is tailored to the ROM_EXT needs.
+// In particular:
+//   - We don't care about the OTP setting for SPX+ in the ROM_EXT.
+//   - We don't care about flash_exec in the ROM_EXT.
+//   - We have a different series of algorithm identifier values to accommodate
+//     hybrid signature schemes.
+OT_WARN_UNUSED_RESULT
+static rom_error_t rom_ext_spx_verify(
+    const sigverify_spx_signature_t *signature, const sigverify_spx_key_t *key,
+    uint32_t key_alg, const void *msg_prefix_1, size_t msg_prefix_1_len,
+    const void *msg_prefix_2, size_t msg_prefix_2_len, const void *msg,
+    size_t msg_len, const hmac_digest_t *digest) {
+  /*
+   * Shares for producing kErrorOk if SPHINCS+ verification succeeds.  The first
+   * three shares are generated using the `sparse-fsm-encode` script while the
+   * last share is
+   * `kErrorOk ^ shares[0] ^ ... ^ shares[2]`.
+   *
+   * Encoding generated with:
+   * $ ./util/design/sparse-fsm-encode.py -d 5 -m 3 -n 32 \
+   *     -s 1069420 --language=c
+   *
+   * Minimum Hamming distance: 14
+   * Maximum Hamming distance: 20
+   * Minimum Hamming weight: 14
+   * Maximum Hamming weight: 16
+   */
+
+  const uint32_t shares[] = {
+      0x11deb806,
+      0x06457f69,
+      0x647f10c4,
+      0x73e4d092,
+  };
+
+  key_alg &= ~(uint32_t)kOwnershipKeyAlgCategoryMask;
+  key_alg |= (uint32_t)kOwnershipKeyAlgCategorySpx;
+
+  sigverify_spx_root_t actual_root;
+  sigverify_spx_root_t expected_root;
+  spx_public_key_root(key->data, expected_root.data);
+  size_t i;
+  for (i = 0; launder32(i) < kSigverifySpxRootNumWords; ++i) {
+    expected_root.data[i] ^= shares[i];
+  }
+
+  switch (key_alg) {
+    case kOwnershipKeyAlgSpxPure:
+      HARDENED_RETURN_IF_ERROR(spx_verify(
+          signature->data, kSpxVerifyPureDomainSep, kSpxVerifyPureDomainSepSize,
+          msg_prefix_1, msg_prefix_1_len, msg_prefix_2, msg_prefix_2_len, msg,
+          msg_len, key->data, actual_root.data));
+      break;
+
+    case kOwnershipKeyAlgSpxPrehash:
+      HARDENED_RETURN_IF_ERROR(
+          spx_verify(signature->data, kSpxVerifyPrehashDomainSep,
+                     kSpxVerifyPrehashDomainSepSize,
+                     /*msg_prefix_2=*/NULL, /*msg_prefix_2_len=*/0,
+                     /*msg_prefix_3=*/NULL, /*msg_prefix_3_len=*/0,
+                     (unsigned char *)digest->digest, sizeof(digest->digest),
+                     key->data, actual_root.data));
+      break;
+    default:
+      return kErrorSigverifyBadSpxConfig;
+  }
+  uint32_t result = 0;
+  for (--i; launder32(i) < kSigverifySpxRootNumWords; --i) {
+    result ^= expected_root.data[i] ^ actual_root.data[i];
+  }
```

Comment:
The algorithm we use in ROM for this comparison is a little more complex and uses an additional mask variable `diff` https://github.com/lowRISC/opentitan/blob/16484156dd156f045ba2c6576c6a21cdca8e3674/sw/device/silicon_creator/lib/sigverify/spx_verify.c#L130

Does it make sense to consider a similar construction here? Or is it appropriate to be less conservative here than in ROM, and perhaps just highlight this to pentesters for evaluation?

Reply:
The use of `diff` seems to be to clobber all of the bits in the result if there is an invalid signature.  I think that makes sense for clobbering the `flash_exec` "password".  We can probably add a similar construction here to clobber the result in a similar way.

Reply:
Done.